types/wlr_input_timestamps_v1.c

```diff
@@ -0,0 +1,383 @@
+	struct wlr_input_timestamps_v1 *input_timestamps =
+		wl_container_of(listener, input_timestamps, seat_destroy);
+
+	destroy(input_timestamps);
+}
+
+static void handle_input_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_input_timestamps_v1 *input_timestamps =
+		wl_container_of(listener, input_timestamps, input_destroy);
+
+	destroy(input_timestamps);
+}
+
+/**
+ * input_timestamps_manager handler functions
+ */
+
+static void manager_handle_resource_destroy(struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void manager_handle_destroy(struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+
+	wlr_log(WLR_DEBUG,
+		"input_timestamps_v1 manager unbound from client %p", client);
+}
+
+static void manager_handle_get_input_timestamps(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id, struct wl_resource *input,
+		enum wlr_input_timestamps_type input_timestamps_type) {
+	struct wlr_input_timestamps_manager_v1 *manager =
+		wlr_input_timestamps_manager_v1_from_resource(resource);
+	assert(manager);
+
+	struct wlr_seat_client *seat_client = NULL;
+	switch (input_timestamps_type) {
+	case WLR_INPUT_TIMESTAMPS_KEYBOARD:
+		seat_client = wlr_seat_client_from_keyboard_resource(input);
+		break;
+	case WLR_INPUT_TIMESTAMPS_POINTER:
+		seat_client = wlr_seat_client_from_pointer_resource(input);
+		break;
+	case WLR_INPUT_TIMESTAMPS_TOUCH:
+		seat_client = wlr_seat_client_from_touch_resource(input);
+		break;
+	}
+
+	struct wlr_input_timestamps_v1 *input_timestamps =
+		calloc(1, sizeof(struct wlr_input_timestamps_v1));
+	if (input_timestamps == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wl_resource *input_timestamps_resource = wl_resource_create(client,
+		&zwp_input_timestamps_v1_interface, wl_resource_get_version(resource), id);
+	if (input_timestamps_resource == NULL) {
+		free(input_timestamps);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	input_timestamps->resource = input_timestamps_resource;
+	input_timestamps->input_resource = input;
+	input_timestamps->seat = seat_client->seat;
+
+	wl_list_insert(&manager->input_timestamps, &input_timestamps->link);
+
+	input_timestamps->input_timestamps_type = input_timestamps_type;
+
+	wl_signal_init(&input_timestamps->events.destroy);
+
+	wl_signal_add(&input_timestamps->seat->events.destroy,
+			&input_timestamps->seat_destroy);
+	input_timestamps->seat_destroy.notify = handle_seat_destroy;
+
+	wl_resource_add_destroy_listener(input_timestamps->input_resource,
+			&input_timestamps->input_destroy);
+	input_timestamps->input_destroy.notify = handle_input_destroy;
+
+	wl_resource_set_implementation(input_timestamps_resource, &impl,
+		input_timestamps, handle_resource_destroy);
+
+	switch (input_timestamps_type) {
+	case WLR_INPUT_TIMESTAMPS_KEYBOARD:
+		wlr_signal_emit_safe(&manager->events.new_keyboard_timestamps,
```

Comment:
Maybe we could have a single signal, instead of 3 separate signals?

Reply:
I wouldn't prefer users checking against the type enum on their own, but upto you.

Reply:
Well, why do we need these signals in the first place? I don't think a new input timestamps object requires any action from the compositor. I'd rather leave this out until there is a clear use-case.

Reply:
Honestly, I can't think of any use-cases except contrived ones like logging or unit testing. They're there mostly to be on par with other protocols.

Reply:
Then let's just remove them